src/cargo/util/auth/mod.rs

```diff
@@ -552,20 +553,37 @@ fn credential_action(
         })?;
         match provider.perform(&registry, &action, &args[1..]) {
             Ok(response) => return Ok(response),
-            Err(cargo_credential::Error::UrlNotSupported) => {}
-            Err(cargo_credential::Error::NotFound) => any_not_found = true,
-            e => {
-                return e.with_context(|| {
-                    format!(
-                        "credential provider `{}` failed action `{action}`",
-                        args.join(" ")
-                    )
-                })
-            }
+            Err(e) => match e.kind() {
+                cargo_credential::ErrorKind::UrlNotSupported => {
+                    if e.as_inner().is_some() {
+                        custom_not_supported.get_or_insert_with(|| {
+                            Err::<(), _>(e)
+                                .with_context(|| {
+                                    format!(
+                                        "credential provider `{}` could not handle the request",
```

Comment:
Could you please add a UI test so we can see the output when there are multiple failed providers?